streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java

```diff
@@ -59,4 +84,49 @@ public SourceNode source(String topic) {
         return stateStoreSuppliers;
     }
 
+
+    private String childrenToString(List<ProcessorNode<?, ?>> children) {
+        if (children == null || children.isEmpty()) {
+            return "";
+        }
+
+        StringBuilder sb = new StringBuilder("children [");
+        for (ProcessorNode child : children) {
+            sb.append(child.name() + ",");
+        }
+        sb.setLength(sb.length() - 1);
+        sb.append("]\n");
+
+        // recursively print children
+        for (ProcessorNode child : children) {
+            sb.append(child.toString());
+            if (!sinkNameToTopic.containsKey(child.name())) {
+                sb.append(childrenToString(child.children()));
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Produces a string representation contain useful information this topology.
+     * This is useful in debugging scenarios.
+     * @return A string representation of this instance.
+     */
+    public String toString() {
+        StringBuilder sb = new StringBuilder("\t\tProcessorTopology:\n");
+
+        // start from sources
+        for (SourceNode source : sourceByTopics.values()) {
+            sb.append("\t\t" + source.name() + ": " + "topics: [");
```

Comment:
Actually I meant to have one more `\t` than `ProcessorTopology` and `Partitions` so that they are a bit more obvious in the hierarchy. I think I was not very clear in my previous comment.